internal/bundler/apko.go

```diff
@@ -0,0 +1,153 @@
+package bundler
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	apko_build "chainguard.dev/apko/pkg/build"
+	apko_oci "chainguard.dev/apko/pkg/build/oci"
+	apko_types "chainguard.dev/apko/pkg/build/types"
+	"chainguard.dev/apko/pkg/tarfs"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+// apko is a bundler that uses the local machine to build the
+// image using apko.
+type apko struct {
+	arch       apko_types.Architecture
+	apkoConfig apko_types.ImageConfiguration
+	ropts      []remote.Option
+}
+
+type ApkoOpt func(*apko) error
+
+func NewApko(opts ...ApkoOpt) (Bundler, error) {
+	a := &apko{
+		arch: apko_types.ParseArchitecture(runtime.GOARCH),
+		apkoConfig: apko_types.ImageConfiguration{
+			Contents: apko_types.ImageContents{
+				RuntimeRepositories: []string{
+					"https://packages.wolfi.dev/os",
+					"https://packages.cgr.dev/extras",
+				},
+				Keyring: []string{
+					"https://packages.wolfi.dev/os/wolfi-signing.rsa.pub",
+					"https://packages.cgr.dev/extras/chainguard-extras.rsa.pub",
+				},
+				Packages: []string{
+					"wolfi-baselayout",
+					"busybox",
+					"bash",
+					"apk-tools",
+				},
+			},
+			Cmd: "/bin/sh -l",
+			Accounts: apko_types.ImageAccounts{
+				RunAs: "0",
+				Users: []apko_types.User{
+					{
+						UserName: "nonroot",
+						UID:      65532,
+						GID:      65532,
+					},
+				},
+				Groups: []apko_types.Group{
+					{
+						GroupName: "nonroot",
+						GID:       65532,
+					},
+				},
+			},
+		},
+	}
+
+	for _, opt := range opts {
+		if err := opt(a); err != nil {
+			return nil, err
+		}
+	}
+
+	return a, nil
+}
+
+func (a *apko) Bundle(ctx context.Context, repo name.Repository, layers ...Layerer) (name.Reference, error) {
+	bopts := []apko_build.Option{
+		apko_build.WithImageConfiguration(a.apkoConfig),
+		apko_build.WithArch(a.arch),
+		// apko will set the cache dir, so don't set it here
+	}
+
+	bc, err := apko_build.New(ctx, tarfs.New(), bopts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create build context: %w", err)
+	}
+
+	bde, err := bc.GetBuildDateEpoch()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get build date epoch: %w", err)
+	}
+
+	_, layer, err := bc.BuildLayer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build layer: %w", err)
+	}
+
+	base, err := apko_oci.BuildImageFromLayer(ctx, empty.Image, layer, bc.ImageConfiguration(), bde, a.arch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build image: %w", err)
+	}
+
+	img, err := appendLayers(base, layers...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to append layers: %w", err)
+	}
+
+	digest, err := img.Digest()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get digest: %w", err)
+	}
+
+	ref := repo.Digest(digest.String())
+
+	pusher, err := remote.NewPusher(a.ropts...)
```

Comment:
It would be nice to make the pusher persistent across Bundle invocations to take advantage of some caching / work deduping

Reply:
whoops, I did all that plumbing to store it at the provider level only to forget to call it